### PR TITLE
SpellStorage doesn't inherit from std::vector anymore

### PR DESCRIPTION
**Pull request by vincent-grosbois**
This is cleaner because now it's impossible to directly modify the underlying std::vector
But it required several refactors in different places, so not sure it's worth it overall

### CODE REVIEW

File: src/fheroes2/spell/spell_storage.h [hunk 1]
```diff
@@ -31,19 +31,38 @@
...
 
-    SpellStorage GetSpells( int lvl = -1 ) const;
-    void Append( const SpellStorage & );
-    void Append( const Spell & );
-    void Append( const BagArtifacts & );
-    void Append( const Artifact & );
-    bool isPresentSpell( const Spell & ) const;
-    bool hasAdventureSpell( const int lvl ) const;
+    // used to allow iteration over this object
+    typedef std::vector<Spell>::const_iterator const_iterator;
```
Comment (qarkai) [inline src/fheroes2/spell/spell_storage.h:40]:
```suggestion
    using const_iterator = std::vector<Spell>::const_iterator;
```

File: src/fheroes2/spell/spell_storage.h [hunk 2]
```diff
@@ -31,19 +31,38 @@
...
+    const_iterator begin() const;
+    const_iterator end() const;
+
+    int Size() const;
```
Comment (qarkai) [inline src/fheroes2/spell/spell_storage.h:44]: Method appears unused.

File: src/fheroes2/spell/spell_book.cpp [hunk 3]
```diff
@@ -457,31 +457,32 @@ void SpellBook::Edit( const HeroBase & hero )
...
+        storage.Append( hero->GetBagArtifacts() );
 
     if ( filter != SpellBook::Filter::ALL ) {
-        res.resize( std::distance( res.begin(), std::remove_if( res.begin(), res.end(), [filter]( const Spell & s ) {
-                                       return ( ( SpellBook::Filter::ADVN == filter ) && s.isCombat() ) || ( ( SpellBook::Filter::CMBT == filter ) && !s.isCombat() );
-                                   } ) ) );
+        storage.spells.resize( std::distance( storage.spells.begin(), std::remove_if( storage.spells.begin(), storage.spells.end(), [filter]( const Spell & s ) {
+                                                  return ( ( SpellBook::Filter::ADVN == filter ) && s.isCombat() )
+                                                         || ( ( SpellBook::Filter::CMBT == filter ) && !s.isCombat() );
+                                              } ) ) );
```
Comment (qarkai) [inline src/fheroes2/spell/spell_book.cpp:472]:
```suggestion
        storage.spells.erase( std::remove_if( storage.spells.begin(), storage.spells.end(), [filter]( const Spell & s ) {
                                                  return ( ( SpellBook::Filter::ADVN == filter ) && s.isCombat() )
                                                         || ( ( SpellBook::Filter::CMBT == filter ) && !s.isCombat() );
                                              } ), storage.spells.end() );
```

File: src/fheroes2/spell/spell_book.cpp [hunk 4]
```diff
@@ -457,31 +457,32 @@ void SpellBook::Edit( const HeroBase & hero )
...
-        SpellStorage::iterator itend = res.end();
-        itend = std::remove( res.begin(), itend, Spell( Spell::TOWNGATE ) );
-        itend = std::remove( res.begin(), itend, Spell( Spell::TOWNPORTAL ) );
-        if ( res.end() != itend )
-            res.resize( std::distance( res.begin(), itend ) );
+        auto itend = storage.spells.end();
+        itend = std::remove( storage.spells.begin(), itend, Spell( Spell::TOWNGATE ) );
+        itend = std::remove( storage.spells.begin(), itend, Spell( Spell::TOWNPORTAL ) );
+        if ( storage.spells.end() != itend )
+            storage.spells.resize( std::distance( storage.spells.begin(), itend ) );
```
Comment (qarkai) [inline src/fheroes2/spell/spell_book.cpp:481]:
```suggestion
        storage.spells.erase( std::remove( storage.spells.begin(), itend, Spell( Spell::TOWNGATE ) ), storage.spells.end() );
        storage.spells.erase( std::remove( storage.spells.begin(), itend, Spell( Spell::TOWNPORTAL ) ), storage.spells.end() );
```

File: src/fheroes2/spell/spell_storage.cpp [hunk 5]
```diff
@@ -29,57 +29,66 @@
...
-    for ( const Spell & sp : st ) {
-        if ( std::find( begin(), end(), sp ) == end() ) {
-            push_back( sp );
-        }
-    }
+    return static_cast<int>( spells.size() );
 }
 
-bool SpellStorage::isPresentSpell( const Spell & spell ) const
+bool SpellStorage::HasSpell( const Spell & spell ) const
```
Comment (qarkai) [inline src/fheroes2/spell/spell_storage.cpp:50]: Please follow https://github.com/ihhub/fheroes2/wiki/Code-style#naming

File: src/fheroes2/spell/spell_storage.cpp [hunk 6]
```diff
@@ -29,57 +29,66 @@
...
-    return end() != std::find( begin(), end(), spell );
+    return std::find( spells.begin(), spells.end(), spell ) != spells.end();
 }
 
-bool SpellStorage::hasAdventureSpell( const int lvl ) const
+bool SpellStorage::HasAdventureSpellAtLevel( const int spellLevel ) const
 {
-    for ( const_iterator it = begin(); it != end(); ++it ) {
-        if ( ( *it ).Level() == lvl && ( *it ).isAdventure() )
+    for ( auto it : spells ) {
```
Comment (qarkai) [inline src/fheroes2/spell/spell_storage.cpp:57]: Please follow https://github.com/ihhub/fheroes2/wiki/Code-style#auto

File: src/fheroes2/spell/spell_storage.h [hunk 7]
```diff
@@ -31,19 +31,38 @@
...
+
+    void Append( const SpellStorage & spellStorage );
+    void Append( const Spell & spell );
+    void Append( const BagArtifacts & artifacts );
+    void Append( const Artifact & artifact );
+
     std::string String( void ) const;
+
+protected:
+    std::vector<Spell> spells;
```
[the remark this answers is not in the document]
Comment (ihhub) [reply src/fheroes2/spell/spell_storage.h:58]: Please name it as `_spells` to indicate that this is a member of the class. Also is there a reason to make this member **protected** over **private**?

File: src/fheroes2/spell/spell_storage.cpp [hunk 8]
```diff
@@ -110,3 +119,23 @@ void SpellStorage::Append( const Artifact & art )
...
 }
+
+std::string SpellStorage::String( void ) const
+{
+    std::ostringstream os;
+
+    for ( auto it : spells )
+        os << it.GetName() << ", ";
+
+    return os.str();
```
Comment (ihhub) [inline src/fheroes2/spell/spell_storage.cpp:130]: We can do this without `ostringstream`:
```
std::string output;

for ( const Spell & spell : spells ) {
    if ( !output.empty() ) {
        output += ", ";
    }
    output += spell.GetName();
}

return output;
```

This version won't produce `, `at the end of the string.

**Review by ihhub — CHANGES_REQUESTED** [review]
Hi @vincent-grosbois , I left several comments here. Would you mind to take a look?

File: src/fheroes2/battle/battle_main.cpp [hunk 9]
```diff
@@ -264,17 +264,15 @@ void Battle::PickupArtifactsAction( HeroBase & hero1, HeroBase & hero2 )
...
-    SpellStorage new_spells;
-    new_spells.reserve( 10 );
+    std::vector<Spell> new_spells;
 
     const Skill::Secondary eagleeye( Skill::Secondary::EAGLEEYE, hero.GetLevelSkill( Skill::Secondary::EAGLEEYE ) );
 
     // filter spells
-    for ( SpellStorage::const_iterator it = spells.begin(); it != spells.end(); ++it ) {
-        const Spell & sp = *it;
+    for ( auto sp : spells ) {
```
[the remark this answers is not in the document]
Comment (vincent-grosbois) [reply src/fheroes2/battle/battle_main.cpp:275]: @ihhub I'm not sure what's the big deal with copying a spell instance, it's just wrapping an int
I'll do it I guess but to me this is really a kind of premature optimization

File: src/fheroes2/spell/spell_storage.h (hunk 7 above)
[the remark this answers is not in the document]
Comment (vincent-grosbois) [reply src/fheroes2/spell/spell_storage.h:58]: it's done so SpellBook class has access to it too

**Comment by ihhub** [comment]
I am closing this pull request due to the low impact on the code and presence of many merge conflicts.

**Pull request closed without merging**